### PR TITLE
Flashlight Decoder for Nemo

### PR DESCRIPTION
**Pull request by trias702**
# What does this PR do ?

Adds Flashlight decoding to Nemo

**Collection**: ASR

# Changelog 
- Added flashlight_decoder.py to collections/asr/modules
- Changed collections/asr/metrics/wer.py and wer_bpe.py to support new setter methods for ctc_decoding object
- Changed collections/asr/parts/submodules/ctc_beam_decoding.py to include all of the new logic for Flashlight decoding which uses the new beam search API
- Added script to scripts/asr_language_modeling/ngram_lm which creates Flashlight format lexicon files from KenLM arpa files

# Usage
Works for lexicon and lexicon-free flashlight decoding

```python
import torch

from nemo.collections.asr.models import ASRModel
from nemo.collections.asr.metrics.wer import CTCDecodingConfig

device = torch.device('cuda')
asr_model = ASRModel.restore_from(...).to(device)
files = ['/path/to/audio1.wav', '/path/to/audio2.wav']

# Lexicon-based
decoding_cfg = CTCDecodingConfig()

decoding_cfg.strategy = "flashlight"
decoding_cfg.beam.search_type = "flashlight"
decoding_cfg.beam.kenlm_path = '/path/to/kenlm.bin'
decoding_cfg.beam.flashlight_cfg.lexicon_path='/path/to/lexicon.lexicon'
decoding_cfg.beam.beam_size = 32
decoding_cfg.beam.beam_alpha = 0.2
decoding_cfg.beam.beam_beta = 0.2
decoding_cfg.beam.flashlight_cfg.beam_size_token = 32
decoding_cfg.beam.flashlight_cfg.beam_threshold = 25.0

# Lexicon-free (for char based languages like Japanese)
decoding_cfg = CTCDecodingConfig()

decoding_cfg.strategy = "flashlight"
decoding_cfg.beam.search_type = "flashlight"
decoding_cfg.beam.kenlm_path = '/path/to/kenlm.bin'
decoding_cfg.beam.flashlight_cfg.lexicon_path=None
decoding_cfg.beam.beam_size = 32
decoding_cfg.beam.beam_alpha = 0.2
decoding_cfg.beam.beam_beta = 0.2
decoding_cfg.beam.flashlight_cfg.beam_size_token = 32
decoding_cfg.beam.flashlight_cfg.beam_threshold = 25.0
decoding_cfg.beam.flashlight_cfg.unit_lm = True


asr_model.change_decoding_strategy(decoding_cfg)

transcripts = asr_model.transcribe(paths2audio_files=files, batch_size=2, return_hypotheses=False, logprobs=False)
```

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?
- [X] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [X] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/modules/flashlight_decoder.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+            raise ModuleNotFoundError(
+                "FlashLightKenLMBeamSearchDecoder requires the installation of flashlight python bindings "
+                "from https://github.com/flashlight/text. Please follow the build instructions there."
+            )
+
+        super().__init__()
+        
+        self.criterion_type = CriterionType.CTC
+        self.nbest = nbest
+        self.tokenizer = self.TokensWrapper(vocabulary, tokenizer)
```
Comment (titu1994) [inline nemo/collections/asr/modules/flashlight_decoder.py:124]: Maybe more natural to call it `self.tokenizer_wrapper` to avoid the notion that its the actual tokenizer

File: nemo/collections/asr/modules/flashlight_decoder.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            
+        if lexicon_path is not None:
+            self.lexicon = load_words(lexicon_path)
+            self.word_dict = create_word_dict(self.lexicon)
+            self.unk_word = self.word_dict.get_index("<unk>")
+    
+            self.lm = KenLM(lm_path, self.word_dict)
+            self.trie = Trie(self.vocab_size, self.silence)
+    
+            start_state = self.lm.start(False)
```
Comment (titu1994) [inline nemo/collections/asr/modules/flashlight_decoder.py:138]: Could you add some comments as to what these steps are doing?

File: nemo/collections/asr/modules/flashlight_decoder.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import torch
+import itertools as it
```
Comment (titu1994) [inline nemo/collections/asr/modules/flashlight_decoder.py:17]: preferred not to use short form for this.

File: nemo/collections/asr/modules/flashlight_decoder.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        for i, token_idx in enumerate(token_idxs):
+            if token_idx == self.blank:
+                continue
+            if i == 0 or token_idx != token_idxs[i-1]:
+                timesteps.append(i)
+        
+        return timesteps
+
+    #@typecheck(ignore_collections=True)
+    @torch.no_grad()
```
Comment (titu1994) [inline nemo/collections/asr/modules/flashlight_decoder.py:222]: Can you add back the types if possible? if not, its fine to keep them commented for now.

File: nemo/collections/asr/modules/flashlight_decoder.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            log_probs = torch.from_numpy(log_probs).float()
+        if log_probs.dim() == 2:
+            log_probs = log_probs.unsqueeze(0)
+        
+        emissions = log_probs.cpu().contiguous()
+        
+        B, T, N = emissions.size()
+        hypos = []
+        for b in range(B):
+            emissions_ptr = emissions.data_ptr() + 4 * b * emissions.stride(0)
```
Comment (titu1994) [inline nemo/collections/asr/modules/flashlight_decoder.py:234]: Can you add some lines of comments to describe what is happening here. useful for future.

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py [hunk 6]
```diff
@@ -191,8 +203,10 @@ def __init__(
             self.search_algorithm = self.default_beam_search
         elif search_type == "pyctcdecode":
             self.search_algorithm = self._pyctcdecode_beam_search
-
+            
             raise NotImplementedError(f"The search type of `pyctcdecode` is currently not supported.\n" f"")
+        elif search_type == "flashlight":
+            self.search_algorithm = self._flashlight_beam_search
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/ctc_beam_decoding.py:209]: No need for _ in function name, it can be public method.

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py [hunk 7]
```diff
@@ -173,6 +184,7 @@ def __init__(
         beam_alpha: float = 1.0,
         beam_beta: float = 0.0,
         kenlm_path: str = None,
+        flashlight: Optional['FlashlightConfig'] = None,
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/ctc_beam_decoding.py:187]: Should be `flashlight_cfg`

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py [hunk 8]
```diff
@@ -416,6 +435,102 @@ def _pyctcdecode_beam_search(
...
+
+                # For subword encoding, NeMo will double encode the subword (multiple tokens) into a
+                # singular unicode id. In doing so, we preserve the semantic of the unicode token, and
+                # compress the size of the final KenLM ARPA / Binary file.
+                # In order to do double encoding, we shift the subword by some token offset.
+                # This step is ignored for character based models.
+                #if self.decoding_type == 'subword':
+                #    pred_token_ids = [ord(c) - self.token_offset for c in candidate[1]]
+                #else:
+                #    pred_token_ids = candidate[1]
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/ctc_beam_decoding.py:514]: Probably the commented stuff can be removed safely, we have the default code for reference if needed in the future

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py [hunk 9]
```diff
@@ -439,6 +554,21 @@ class PyCTCDecodeConfig:
     hotword_weight: float = 10.0
 
 
+@dataclass
+class FlashlightConfig:
+    #kenlm_path: Optional[str] = None
+    lexicon_path: Optional[str] = None
+    nbest: int = 1
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/ctc_beam_decoding.py:561]: isnt nbest same as beam_size? Can we drop it and auto fill from the beam_config.beam_size ?

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py [hunk 10]
```diff
@@ -439,6 +554,21 @@ class PyCTCDecodeConfig:
...
+    #beam_size: int = 16
+    beam_size_token: int = 16
+    beam_threshold: float = 20.0
+    #lm_weight: float = 0.2
+    #word_score: float = 0.2
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/ctc_beam_decoding.py:566]: Are these two args not supported? If so, could you remove them?

File: nemo/collections/asr/parts/utils/transcribe_utils.py [hunk 11]
```diff
@@ -289,7 +289,7 @@ def write_transcription(
     else:
         pred_text_attr_name = 'pred_text'
 
-    with open(cfg.output_filename, 'w', encoding='utf-8') as f:
+    with open(cfg.output_filename, 'w', encoding='utf-8', newline='\n') as f:
```
Comment (titu1994) [inline nemo/collections/asr/parts/utils/transcribe_utils.py:292]: Hmm is this newline part needed?

**Review by titu1994 — COMMENTED** [review]
Fantastic work ! Mostly ready to merge, just minor changes required and formatting.

File: nemo/collections/asr/parts/utils/transcribe_utils.py (hunk 11 above)
Comment (trias702) [reply nemo/collections/asr/parts/utils/transcribe_utils.py:292]: So that it works correctly on Windows, which is my primary Nemo working environment. newline='\n' sets the file to use UNIX EOL, so this change is harmless on Unix, but on Windows is necessary to output UNIX EOL files, which makes it in parity with Nemo on unix

File: nemo/collections/asr/modules/flashlight_decoder.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        }
+
+    @property
+    def output_types(self):
+        """Returns definitions of module output ports.
+        """
+        return {"hypos": NeuralType(('B'), PredictionsType())}
+    '''
+    
+    class TokensWrapper:
```
[the remark this answers is not in the document]
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py:44]: I made it inner because it's just a special helper class just for the flashlight_decoder and is necessary for it to work, and it has no use outside of that. I think it's better to keep it as an inner class for this reason

**Comment by XuesongYang** [comment]
@trias702 could you please run `pre-commit run --all-files`  locally, and push the changes again?

File: nemo/collections/asr/modules/flashlight_decoder.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/asr/modules/flashlight_decoder.py:44]: Pickle has an issue with inner classes, and we may need to eventually support pickle so lets extract it out as a private class in this file.

[the remark this answers is not in the document]
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py:44]: Pickle doesn't work for Flashlight regardless, I have already tried, with or without the inner classes, you can't pickle this module for multiprocessing for faster decoding, I already tried

File: nemo/collections/asr/modules/flashlight_decoder.py [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+        if log_probs.dim() == 2:
+            log_probs = log_probs.unsqueeze(0)
+
+        emissions = log_probs.cpu().contiguous()
+
+        B, T, N = emissions.size()
+        hypos = []
+        for b in range(B):
+            emissions_ptr = emissions.data_ptr() + 4 * b * emissions.stride(0)
+            results = self.decoder.decode(emissions_ptr, T, N)
```
[the remark this answers is not in the document]
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py:225]: Yes, that is how the upstream model returns the information, with some sort of padding (I presume it's the blank or silence token, you can see in the model code itself). Since we specify both of those tokens to the flashlight module (see above in the code), they are naturally removed anyway, so it doesn't bother the decoding at all that they're all the same length and padded.

File: nemo/collections/asr/modules/flashlight_decoder.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py]: Fixed

File: nemo/collections/asr/modules/flashlight_decoder.py (hunk 1 above)
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py]: Done

File: nemo/collections/asr/modules/flashlight_decoder.py (hunk 2 above)
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py:141]: Done

File: nemo/collections/asr/modules/flashlight_decoder.py (hunk 3 above)
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py]: Fixed

File: nemo/collections/asr/modules/flashlight_decoder.py (hunk 4 above)
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py:219]: It's not possible as we discussed in our meeting, will leave them uncommented for now

File: nemo/collections/asr/modules/flashlight_decoder.py (hunk 5 above)
Comment (trias702) [reply nemo/collections/asr/modules/flashlight_decoder.py:235]: Done

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py (hunk 6 above)
Comment (trias702) [reply nemo/collections/asr/parts/submodules/ctc_beam_decoding.py]: Fixed

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py (hunk 7 above)
Comment (trias702) [reply nemo/collections/asr/parts/submodules/ctc_beam_decoding.py]: Fixed

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py (hunk 8 above)
Comment (trias702) [reply nemo/collections/asr/parts/submodules/ctc_beam_decoding.py]: Done

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py (hunk 9 above)
Comment (trias702) [reply nemo/collections/asr/parts/submodules/ctc_beam_decoding.py]: I have removed nbest, it wasn't needed at all, sorry about the confusion

File: nemo/collections/asr/parts/submodules/ctc_beam_decoding.py (hunk 10 above)
Comment (trias702) [reply nemo/collections/asr/parts/submodules/ctc_beam_decoding.py]: Removed

**Review by titu1994 — APPROVED** [review]
Looks fantastic, thanks !
